File: api/irfaran/stats.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""What the archive adds up to.

Everything here is derived - not one number is stored as truth - so all of it
can be thrown away and recomputed. What it is not is cheap: measuring how much
ground has been cleared means reading every fog blob in the archive and
counting bits, which is three seconds today and will be more later. So it is
computed on demand, written down with a fingerprint of the archive it was
computed from, and handed back untouched until that fingerprint changes.

Two things about the area that are easy to get wrong and were:

  **Mercator is not a map of areas.** A pixel near the poles covers a fraction
  of the ground a pixel at the equator does - at zoom 14, a tile is 5.97 km2 at
  the equator and 0.72 km2 at the latitude of Tromso. Counting cleared pixels
  and calling the answer area would flatter anybody who has been north, and by
  a factor of eight. Every pixel here is weighted by the true ground area of
  its row. See geo.tile_row_area_m2.

  **The same street in two years is one piece of ground.** Fog is stored per
  source and per layer, so a commute walked in 2019 and again in 2024 has two
  blobs covering identical pixels. They are unioned per tile before anything is
  counted, or the figure would grow every year without anybody going anywhere
  new.

Water counts. The denominator is the whole planet, ocean included, because a
ferry crossing is as much a place you have been as a footpath - and because
excluding water would need a land mask nobody has.
"""

from __future__ import annotations

import json
import sqlite3
from typing import Iterator

import numpy as np

from irfaran import geo, raster

#: Where the last computed answer is kept, with what it was computed from.
CACHE_SETTING = "stats_cache"
FINGERPRINT_SETTING = "stats_fingerprint"

#: Sources that arrive without anybody asking, as opposed to a file somebody
#: dropped in or a line somebody drew.
LIVE_SOURCES = ("overland", "owntracks", "ha")


def fingerprint(conn: sqlite3.Connection) -> str:
    """Something cheap that changes whenever any of the numbers could have.

    Three counts and a high-water mark, all of them index reads. The point is
    that asking "is the cached answer still true?" must not cost anything like
    what computing the answer costs, or the cache is pointless.
    """
    events = conn.execute(
        "SELECT count(*) AS n, COALESCE(max(id), 0) AS top FROM events"
    ).fetchone()
    blobs = conn.execute("SELECT count(*) AS n FROM blobs").fetchone()["n"]
    places = conn.execute("SELECT count(*) AS n FROM places").fetchone()["n"]
    return f"{events['n']}:{events['top']}:{blobs}:{places}"


# ------------------------------------------------------------------- ground


def cleared_tiles(conn: sqlite3.Connection) -> Iterator[tuple[int, int, np.ndarray]]:
    """Every z14 tile that holds fog, as one mask per tile.

    Blobs arrive grouped by tile so that the union can be built and released
    one tile at a time. Holding the whole archive's fog in memory at once is
    three hundred megabytes of masks for no reason.
    """
    rows = conn.execute(
        "SELECT x, y, source, layer, data FROM blobs WHERE kind = 'fog' "
        "ORDER BY x, y"
    )
    current: tuple[int, int] | None = None
    mask: np.ndarray | None = None

    for row in rows:
        here = (int(row["x"]), int(row["y"]))
        if here != current:
            if mask is not None and current is not None:
                yield current[0], current[1], mask
            current, mask = here, None

        blob = raster.decode(
            row["data"], "fog", row["source"], row["layer"], here[0], here[1]
        )
        # The union, not the sum: ground walked in two different years is one
        # piece of ground.
        mask = (blob > 0) if mask is None else (mask | (blob > 0))

    if mask is not None and current is not None:
        yield current[0], current[1], mask


def ground(conn: sqlite3.Connection) -> dict[str, object]:
    """How much of the planet has had its fog cleared."""
    square_metres = 0.0
    tiles = 0

    for _, tile_y, mask in cleared_tiles(conn):
        lit = int(np.count_nonzero(mask))
        if not lit:
            continue
        tiles += 1
        square_metres += lit * geo.tile_pixel_area_m2(tile_y)

    return {
        "square_metres": square_metres,
        "square_km": square_metres / 1e6,
        "tiles": tiles,
        # Of the whole planet rather than of the tile grid. The grid stops at
        # the latitudes Mercator can draw, so it is 0.37% short of a planet -
        # but that missing sliver is the two polar caps, and the question was
        # what fraction of the world has been seen.
        "percent_of_planet": square_metres / geo.EARTH_SURFACE_M2 * 100.0,
    }


# ------------------------------------------------------------------- counting


def routes(conn: sqlite3.Connection) -> dict[str, object]:
    """How many routes there are, and where they came from.

    Workouts are one bucket rather than two, and that is the data's doing: a
    file dropped in under Import and an activity a tracker fetched are both
    written as source `workout` with the same shape, deliberately, so that an
    activity already imported by hand is recognised rather than drawn twice.
    Nothing in the log distinguishes them afterwards.
    """
    counts = {
        row["source"]: int(row["n"])
        for row in conn.execute(
            "SELECT source, count(*) AS n FROM events "
            "WHERE op = 'add' GROUP BY source"
        )
    }

    live = sum(counts.get(source, 0) for source in LIVE_SOURCES)
    workouts = counts.get("workout", 0)
    drawn = counts.get("manual", 0)

    # What a person would call a route, as well as what the log calls one: a
    # long ride arrives as one event per continuous stretch, so a bare count
    # of rows says more journeys than anybody made.
    names: set[str] = set()
    for row in conn.execute(
        "SELECT meta FROM events WHERE op = 'add' AND meta IS NOT NULL"
    ):
        meta = json.loads(row["meta"])
        if meta.get("track"):
            names.add(str(meta["track"]))

    return {
        "total": live + workouts + drawn,
        "workouts": workouts,
        "live": live,
        "drawn": drawn,
        "by_source": counts,
        "named_journeys": len(names),
    }


def marks(conn: sqlite3.Connection) -> dict[str, object]:
    """The other things drawn on the map by hand."""
    one = lambda sql: int(conn.execute(sql).fetchone()[0])  # noqa: E731
    return {
        "pins": one("SELECT count(*) FROM places"),
        "labels": one("SELECT count(*) FROM labels"),
        "folders": one("SELECT count(*) FROM folders"),
        "people": one("SELECT count(*) FROM people"),
        "revealed": one("SELECT count(*) FROM events WHERE op = 'reveal'"),
        "refogged": one("SELECT count(*) FROM events WHERE op = 'erase'"),
    }


def points(conn: sqlite3.Connection) -> int:
    """How many positions the archive is built from.

    Read out of each event's meta rather than by counting coordinates, which
    would mean parsing 28 MB of geometry to answer a question nobody needs to
    the digit. It is what each source reported, so a track imported twice in
    two formats counts twice - which is the honest answer to "how many points
    are on the map" anyway.
    """
    total = 0
    for row in conn.execute(
        "SELECT meta FROM events WHERE op = 'add' AND meta IS NOT NULL"
    ):
        total += int(json.loads(row["meta"]).get("fixes") or 0)
    return total


def years(conn: sqlite3.Connection) -> dict[str, object]:
    """Which time layers have anything in them."""
    layers = [
        str(row["layer"])
        for row in conn.execute(
            "SELECT DISTINCT layer FROM blobs WHERE kind IN ('fog', 'trail') "
            "ORDER BY layer"
        )
    ]
    dated = [layer for layer in layers if layer.isdigit()]
    return {
        "years": dated,
        "count": len(dated),
        "first": dated[0] if dated else None,
        "last": dated[-1] if dated else None,
        "undated": "prehistory" in layers,
    }


# -------------------------------------------------------------------- the lot


def compute(conn: sqlite3.Connection) -> dict[str, object]:
    """Every figure, from scratch. Seconds, not milliseconds."""
    return {
        "ground": ground(conn),
        "routes": routes(conn),
        "points": points(conn),
        "marks": marks(conn),
        "time": years(conn),
    }


def overview(
    conn: sqlite3.Connection, *, refresh: bool = False
) -> dict[str, object]:
    """The figures, computed if the archive has moved since last time.

    The same shape of mistake as the render status recomputing its job count
    on every poll would be fatal here - reading every fog blob is three seconds
    on this archive - so the answer is written down beside a fingerprint of
    what produced it, and only recomputed when that changes.
    """
    mark = fingerprint(conn)
    if not refresh:
        stored = conn.execute(
            "SELECT key, value FROM settings WHERE key IN (?, ?)",
            (CACHE_SETTING, FINGERPRINT_SETTING),
        ).fetchall()
        held = {row["key"]: row["value"] for row in stored}
        if held.get(FINGERPRINT_SETTING) == mark and held.get(CACHE_SETTING):
            try:
                return {**json.loads(held[CACHE_SETTING]), "cached": True}
            except json.JSONDecodeError:
                pass

    figures = compute(conn)
    for key, value in (
        (CACHE_SETTING, json.dumps(figures, separators=(",", ":"))),
        (FINGERPRINT_SETTING, mark),
    ):
        conn.execute(
            "INSERT INTO settings (key, value) VALUES (?, ?) "
            "ON CONFLICT(key) DO UPDATE SET value = excluded.value",
            (key, value),
        )
    return {**figures, "cached": False}

File: web/src/stats.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// What the archive adds up to.
//
// Deliberately not polled. Measuring the cleared ground reads every fog blob
// in the archive - under a second today and more later - so the server caches
// it against a fingerprint and this asks once, when the tab is opened.

import { apiGet } from './api'
import { shownUiTheme } from './theme'
import { element } from './ui'

interface Ground {
  square_km: number
  tiles: number
  percent_of_planet: number
}

interface Routes {
  total: number
  workouts: number
  live: number
  drawn: number
  named_journeys: number
  by_source: Record<string, number>
}

interface Marks {
  pins: number
  labels: number
  folders: number
  people: number
  revealed: number
  refogged: number
}

interface Time {
  years: string[]
  count: number
  first: string | null
  last: string | null
  undated: boolean
}

interface Visited {
  code: string
  name: string
  square_km: number
  square_metres: number
  of_country_percent: number
  country_square_km: number
  pins: number
  because: string
}

interface Countries {
  available: boolean
  why?: string
  countries: Visited[]
  marginal: Visited[]
  at_sea_square_km: number
}

interface Overview {
  ground: Ground
  countries: Countries
  routes: Routes
  points: number
  marks: Marks
  time: Time
  cached: boolean
}

const count = (value: number) => value.toLocaleString()

/**
 * How wide the world is drawn, in pixels.
 *
 * It runs the full width of the sheet rather than half of it, so 960 was being
 * stretched. 1440 covers the widest the sheet gets, and costs 23 KB against 12
 * - still a thumbnail beside the 2.26 MB of polygons it is made from.
 */
const WORLD_WIDTH = 1440

/** Square kilometres, to a sensible number of digits for how big it is. */
export function formatArea(squareKm: number): string {
  if (squareKm >= 10_000) return `${Math.round(squareKm).toLocaleString()} km²`
  if (squareKm >= 100) return `${squareKm.toFixed(1)} km²`
  if (squareKm >= 1) return `${squareKm.toFixed(2)} km²`
  return `${Math.round(squareKm * 1e6).toLocaleString()} m²`
}

/**
 * The share of the planet, said twice.
 *
 * As a percentage it is five leading zeros and means nothing to anybody. As
 * "one part in two million" it means something immediately, which is the point
 * of showing it at all: not to flatter the number but to say how big the world
 * is.
 */
export function formatShare(percent: number): string {
  if (percent <= 0) return 'none of it yet'
  const digits = percent < 0.001 ? 7 : percent < 1 ? 4 : 2
  const share = `${percent.toFixed(digits)}%`
  const oneIn = Math.round(100 / percent)
  return `${share} — about one part in ${count(oneIn)}`
}

export class Stats {
  private asked = false

  /** Told when the page is showing; fetches once, the first time. */
  watch(showing: boolean): void {
    if (!showing) return
    // Every time it is opened, because the theme can have changed since the
    // figures were fetched and the world is drawn in one of two palettes.
    // Costs nothing when it has not: same URL, and the browser has it.
    this.paintWorld()
    if (this.asked) return
    this.asked = true
    void this.load()
  }

  /**
   * The little world at the bottom.
   *
   * An image the server draws rather than shapes sent here: the polygons are
   * 548,471 vertices and this is eight kilobytes. The URL carries the theme
   * because the picture is drawn in it, and the countries because that is what
   * makes it a different picture - so the browser caches each answer and asks
   * again only when the answer has moved.
   */
  private paintWorld(figures?: Overview): void {
    const world = element<HTMLImageElement>('stat-world')
    const theme = shownUiTheme()
    const next = `/api/stats/world.png?theme=${theme}&width=${WORLD_WIDTH}`
    if (!world.src.endsWith(next)) world.src = next

    if (!figures) return
    const seen = figures.countries.countries.length
    world.alt = figures.countries.available
      ? `A world map with ${seen} ${seen === 1 ? 'country' : 'countries'} ` +
        'marked as visited.'
      : 'A world map. The country borders are not available.'
  }

  wire(): void {
    element('stats-refresh').addEventListener('click', () => void this.load(true))
  }

  async load(refresh = false): Promise<void> {
    const button = element<HTMLButtonElement>('stats-refresh')
    const state = element('stats-state')
    button.disabled = true
    state.textContent = refresh ? 'Counting…' : 'Reading…'
    state.hidden = false

    try {
      const figures = await apiGet<Overview>(
        refresh ? '/api/stats?refresh=true' : '/api/stats',
      )
      this.paint(figures)
      state.hidden = true
    } catch {
      state.textContent =
        'Could not read the figures. The browser console has the detail.'
      state.hidden = false
    } finally {
      button.disabled = false
    }
  }

  /** A share of a country, at whatever precision makes it a number. */
  private static share(percent: number): string {
    if (percent >= 1) return `${percent.toFixed(1)}%`
    if (percent >= 0.01) return `${percent.toFixed(2)}%`
    return `${percent.toFixed(4)}%`
  }

  private paintCountries(visited: Countries): void {
    const host = element('stat-country-list')
    host.textContent = ''

    if (!visited.available) {
      element('stat-countries').textContent = '—'
      element('stat-at-sea').textContent =
        visited.why ?? 'The country borders are not available.'
      return
    }

    element('stat-countries').textContent = String(visited.countries.length)

    for (const row of visited.countries) {
      const line = document.createElement('div')
      line.className = 'country-row'

      const name = document.createElement('span')
      name.textContent = row.name

      const area = document.createElement('span')
      area.className = 'country-area'
      area.textContent = formatArea(row.square_km)

      const share = document.createElement('span')
      share.className = 'country-share'
      share.textContent = Stats.share(row.of_country_percent)
      share.title =
        `${formatArea(row.square_km)} of ${formatArea(row.country_square_km)}` +
        (row.pins ? ` · ${row.pins} pin${row.pins === 1 ? '' : 's'}` : '')

      line.append(name, area, share)
      host.append(line)
    }

    const marginal = element('stat-country-marginal')
    marginal.hidden = visited.marginal.length === 0
    marginal.textContent = visited.marginal.length
      ? 'Too little to call a visit, and shown rather than dropped — this is ' +
        'where an approximate border would turn up: ' +
        visited.marginal.map((row) => row.name).join(', ')
      : ''

    element('stat-at-sea').textContent =
      `${formatArea(visited.at_sea_square_km)} in no country at all — at sea, ` +
      'or past where the borders reach'
  }

  private paint(figures: Overview): void {
    const { ground, routes, marks, time } = figures
    this.paintCountries(figures.countries)
    this.paintWorld(figures)

    element('stat-area').textContent = formatArea(ground.square_km)
    element('stat-share').textContent = formatShare(ground.percent_of_planet)
    element('stat-tiles').textContent =
      `${count(ground.tiles)} squares of the map have been touched`

    element('stat-routes').textContent = count(routes.total)
    element('stat-routes-workouts').textContent = count(routes.workouts)
    element('stat-routes-drawn').textContent = count(routes.drawn)
    element('stat-routes-live').textContent = count(routes.live)
    element('stat-journeys').textContent = count(routes.named_journeys)

    element('stat-points').textContent = count(figures.points)

    element('stat-pins').textContent = count(marks.pins)
    element('stat-labels').textContent = count(marks.labels)
    element('stat-folders').textContent = count(marks.folders)
    element('stat-people').textContent = count(marks.people)
    element('stat-revealed').textContent = count(marks.revealed)
    element('stat-refogged').textContent = count(marks.refogged)

    element('stat-years').textContent = time.count
      ? `${time.first} to ${time.last}`
      : 'nothing dated yet'
    element('stat-years-count').textContent = time.count
      ? `${count(time.count)} ${time.count === 1 ? 'year' : 'years'}` +
        (time.undated ? ', and everything before records' : '')
      : time.undated
        ? 'everything is undated'
        : ''
  }
}
